### PR TITLE
ECDH using libsecp256k1

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

**Comment by arkpar** [comment]
Should be merged together:
https://github.com/ethereum/libweb3core/pull/15
https://github.com/ethereum/webthree-helpers/pull/36
https://github.com/ethereum/libethereum/pull/36

File: libdevcrypto/Common.cpp [hunk 1]
```diff
@@ -40,23 +41,18 @@ using namespace dev;
...
-	secp256k1_context_t* ctx;
-	operator secp256k1_context_t const*() const { return ctx; }
-};
-static Secp256k1Context s_secp256k1;
+std::unique_ptr<secp256k1_context, void(*)(secp256k1_context*)> s_secp256k1(secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY), [](secp256k1_context* _c){ secp256k1_context_destroy(_c); });
 #endif
 
 static Secp256k1PP s_secp256k1pp;
 
+static h256 const c_curveN("0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141");
```
Comment (chfast) [inline libdevcrypto/Common.cpp:49]: Place it inside the function.

File: libdevcrypto/Common.cpp [hunk 2]
```diff
@@ -73,11 +69,14 @@ Address dev::ZeroAddress = Address();
...
 #ifdef ETH_HAVE_SECP256K1
-	bytes o(65);
-	int pubkeylen;
-	if (!secp256k1_ec_pubkey_create(s_secp256k1, o.data(), &pubkeylen, _secret.data(), false))
+	secp256k1_pubkey pub;
+	if (!secp256k1_ec_pubkey_create(s_secp256k1.get(), &pub, _secret.data()))
+		return Public();
+	size_t outlen = 0;
+	byte out[65];
+	if (!secp256k1_ec_pubkey_serialize(s_secp256k1.get(), out, &outlen, &pub, 0))
```
Comment (chfast) [inline libdevcrypto/Common.cpp]: Why `outlen` is ignored?

File: libdevcrypto/Common.cpp (hunk 1 above)
Comment (arkpar) [inline libdevcrypto/Common.cpp:49]: Function static initialization is still not thread safe in msvc

Comment (chfast) [inline libdevcrypto/Common.cpp:49]: So use local variable.

It does not look to be secp256k1 related.

Comment (arkpar) [inline libdevcrypto/Common.cpp:49]: The point of this change is to avoid creating h256 from hex string on each function call

Comment (chfast) [inline libdevcrypto/Common.cpp:49]: I know that.

Can you add "FIXME" comment to move it back into the function after switch to VS 2015?

File: libdevcrypto/ECDHE.cpp [hunk 3]
```diff
@@ -41,6 +60,5 @@ void ECDHE::agree(Public const& _remote, Secret& o_sharedSecret) const
 		BOOST_THROW_EXCEPTION(InvalidState());
 	
 	m_remoteEphemeral = _remote;
-	s_secp256k1.agree(m_ephemeral.sec(), m_remoteEphemeral, o_sharedSecret);
+	dev::crypto::ecdh::agree(m_ephemeral.sec(), m_remoteEphemeral, o_sharedSecret);
```
Comment (chfast) [inline libdevcrypto/ECDHE.cpp:63]: Prefix dev::crypto:: not needed.

File: libdevcrypto/ECDHE.cpp [hunk 4]
```diff
@@ -21,17 +21,36 @@
...
+#if ETH_HAVE_SECP256K1
+#include <secp256k1/include/ext.h>
+#endif
 
 using namespace std;
 using namespace dev;
 using namespace dev::crypto;
 
+#if ETH_HAVE_SECP256K1
+extern std::unique_ptr<secp256k1_context, void(*)(secp256k1_context*)> s_secp256k1;
```
Comment (chfast) [inline libdevcrypto/ECDHE.cpp:35]: Place that in a header if you want to share the context.

**Comment by chfast** [comment]
Small changes needed.
I'm not able to verify security logic. Hope it is covered by unit tests.

**Comment by arkpar** [comment]
Closing this per @subtly's request

**Pull request closed without merging**